### PR TITLE
Provide setting to define a last-will message

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

**Comment by floitsch** [comment]
Most of the commits in this PR come from https://github.com/toitware/mqtt/pull/6.
I just added the finishing touches.

File: src/packets.toit [hunk 1]
```diff
@@ -5,12 +5,38 @@
...
 
 interface PacketIDAck:
   packet_id -> int
 
+class LastWillConfig:
+  retain/bool
+  qos/int
+  topic/string
+  payload/ByteArray
+
```
Comment (kasperl) [inline src/packets.toit:20]: Drop one newline.

File: src/packets.toit [hunk 2]
```diff
@@ -5,12 +5,38 @@
...
 import reader
+import .client show Client  // For toitdoc reference.
 
 import .topic_filter
 
 interface PacketIDAck:
   packet_id -> int
 
+class LastWillConfig:
+  retain/bool
```
Comment (kasperl) [inline src/packets.toit:16]: Does our style guide suggest having a space before / for these types?

File: src/packets.toit [hunk 3]
```diff
@@ -5,12 +5,38 @@
...
+    is eventually sent.
+  */
+  constructor .topic .payload --.qos --.retain=false:
+    if not 0 <= qos <= 2: throw "INVALID_QOS"
+
+  /**
+  Variant of $LastWillConfig.constructor.
+  Takes a string ($message) as payload, instead of a ByteArray.
+  */
+  constructor.from_string .topic message/string --.qos --.retain=false:
```
Comment (kasperl) [inline src/packets.toit:36]: Not sure 'from_string' is great here. It is just the message that is from a string, not the full config.

File: src/packets.toit [hunk 4]
```diff
@@ -5,12 +5,38 @@
...
 import binary
 import reader
+import .client show Client  // For toitdoc reference.
 
 import .topic_filter
 
 interface PacketIDAck:
   packet_id -> int
 
+class LastWillConfig:
```
Comment (kasperl) [inline src/packets.toit:15]: I wonder if having Config as part of this name adds anything. I'd suggest removing it.

**Review by kasperl — APPROVED** [review]
Looks good to me with a few suggestions.

File: src/packets.toit (hunk 4 above)
Comment (floitsch) [reply src/packets.toit]: removed.
done.

File: src/packets.toit (hunk 2 above)
Comment (floitsch) [reply src/packets.toit:16]: yes, but the whole package was written without spaces...

File: src/packets.toit (hunk 1 above)
Comment (floitsch) [reply src/packets.toit:20]: done.

File: src/packets.toit (hunk 3 above)
Comment (floitsch) [reply src/packets.toit]: Removed it.
The `publish` function only takes byte-arrays too.

Hopefully with the new io library things should get nicer.
done.

**Pull request merged**